### ISSUE TITLE
Demo doesn't work

### ISSUE THREAD

**Issue by OmgImAlexis** [blevesearch/bleve#600]
Your demo isn't working as the url it's using for search is just redirecting to a page returning a 404.

https://bleve-search.appspot.com/api/search.bleve/_search

**Comment by mschoch** [comment]
Search on the blevesearch website seems to be working fine.  Can you explain what you did to get the error?

The request handler you linked to only accepts POST requests, not GET requests:

```
➜  ~ curl http://bleve-search.appspot.com/api/search.bleve/_search
404 page not found
➜  ~ curl http://bleve-search.appspot.com/api/search.bleve/_search -d '{"query":{"match":"water"}}'
{"status":{"total":1,"failed":0,"successful":1},"request":{"query":{"match":"water","prefix_length":0,"fuzziness":0},"size":10,"from":0,"highlight":null,"fields":null,"facets":null,"explain":false,"sort":["-_score"]},"hits":[{"index":"search.bleve","id":"/docs/Query-String-Query/","score":0.43571167406343025,"locations":{"content":{"water":[{"pos":47,"start":284,"end":289,"array_positions":null},{"pos":56,"start":330,"end":335,"array_positions":null},{"pos":118,"start":708,"end":713,"array_positions":null},{"pos":212,"start":1291,"end":1296,"array_positions":null},{"pos":288,"start":1766,"end":1771,"array_positions":null}]}},"sort":["_score"]},{"index":"search.bleve","id":"/docs/Text-Analysis/","score":0.34884268864071494,"locations":{"content":{"water":[{"pos":134,"start":751,"end":756,"array_positions":null},{"pos":150,"start":850,"end":855,"array_positions":null}]}},"sort":["_score"]}],"total_hits":2,"max_score":0.43571167406343025,"took":167083,"facets":{}}
```

**Comment by OmgImAlexis** [comment]
I simply went to the demo page and it shows "network error".

The network error is from a redirect.
```
XMLHttpRequest cannot load http://bleve-search.appspot.com/api/search.bleve/_search. Response for preflight is invalid (redirect)
```

**Comment by mschoch** [comment]
Odd, when I click the link you provide I see the following:

![demo](https://cloud.githubusercontent.com/assets/298143/26581977/187f8c28-450d-11e7-843d-d99932455399.png)

**Comment by mschoch** [comment]
What browser are you using, and is there any network proxy involved that you know of?

**Comment by OmgImAlexis** [comment]
Chrome Version 58.0.3029.110 (64-bit) on MacOS Sierra 10.12.3 (16D32).

No network proxy or VPN. The HTTP error is valid though as a preflight request(OPTION) should not contain a redirect.

**Comment by mschoch** [comment]
I have no doubt the error is valid, I just don't know anything about it.

It's a very simple Go HTTP handler with some sort of CORS wrapper we copy/pasted from someone else to allow the browser access it from another domain.

The HTTP handlers are setup here:

https://github.com/blevesearch/bleve-hosted/blob/master/bleve.go#L31-L45

And the CORS wrapper is here:

https://github.com/blevesearch/bleve-hosted/blob/master/cors.go

**Comment by mschoch** [comment]
Also, my browser reports the identical version.

**Comment by mschoch** [comment]
From searching around it seems like its related to CORS, but again I don't really know what the solution is.  If someone wants to propose something I'll try it.

**Comment by OmgImAlexis** [comment]
This is really odd using incognito mode the site works fine. I have all my extensions enabled for both normal browsing and incognito mode so I'm really not sure what the issue is.

**Comment by OmgImAlexis** [comment]
Even if I disable all my extensions I'm still only getting this error in normal browsing mode.

**Comment by OmgImAlexis** [comment]
So it looks like this may be failing because of a caching issue with preflight and CORS.
3XX redirects are only allowed if the request does NOT require a preflight check. Since the library you're using for the request uses OPTIONS it's causing the issue. 

https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0

**Comment by mschoch** [comment]
Can you look in network section of the chrome developer tools and report what you see.  I filtered by XHR and see the following:

![dev](https://cloud.githubusercontent.com/assets/298143/26583450/467fd21c-4513-11e7-83bc-6462156ed7a4.jpg)

**Comment by OmgImAlexis** [comment]
![screen shot 2017-05-30 at 10 20 56 pm](https://cloud.githubusercontent.com/assets/6525926/26583956/4c2be8e0-4586-11e7-9f49-411a3313a360.png)

**Comment by mschoch** [comment]
One difference I see is that your OPTIONS request is sent with `Origin: null`, whereas mine is sent with `Origin: http://www.blevesearch.com/`.

That difference would seem to affect the logic executed here:

https://github.com/blevesearch/bleve-hosted/blob/master/cors.go#L24

A little more googling has led to this:

https://stackoverflow.com/questions/8456538/origin-null-is-not-allowed-by-access-control-allow-origin

I don't really know why the Origin header is null in your case.

**Comment by OmgImAlexis** [comment]
That would be one of multiple privacy extensions I have installed. The main problem with that is the CORS request is cached meaning I'd have to clear my whole cache in order to view the site. Should I close this and you can just mark it as "won't fix" since it's less of an issue with the site and more of an issue with the fact that my extension is messing with the headers?

**Comment by mschoch** [comment]
Let's not close it yet.  I think my understanding was a bit wrong, as null is just a string in this case, so it seems that should be unrelated to the execution path followed.

The intent of our CORS handler is to basically ignore the origin, and allow the browser to send the request.  So, if possible I'd like to get it to work in your case as well.

Can you share the name of the extension you're using?

**Comment by OmgImAlexis** [comment]
If that's the case then there's no need to check the origin as you are in [cors.go](https://github.com/blevesearch/bleve-hosted/blob/master/cors.go#L24).

The extension I'm using is ScriptSafe with all the privacy options enabled.

**Comment by mschoch** [comment]
Yes, I agree that origin check could be removed, but I now think its unrelated to the problem.  I'm pretty sure origin = "null" in that case, and is still executed.  Looking again, the response shown does contain the headers that we add, even stranger it shows `Access-Control-Allow-Origin: http://www.blevesearch.com/` which suggests on the server side, we did see the correct `Origin` header.

Assuming for the moment that value of the `Origin` header is unrelated.  I still don't understand where the redirect is coming from.

**Comment by mschoch** [comment]
I've tried used ScriptSafe to reproduce the problem here, but it seems that by default it won't even execute anything.  Turning on either "allow" or "trust" seems to go far, as the OPTIONS request comes back with 200.  Not sure if I just need to tweak some settings, or perhaps this extension isn't directly related as originally thought.
